### ISSUE TITLE
Seemingly random CI failures of ``tests/test_intl.py::test_gettext_dont_rebuild_mo`` with Windows

### ISSUE THREAD

**Issue by jfbu** [sphinx-doc/sphinx#11232]
### Describe the bug

I am (edit: *not*) competent to understand seemingly occasional:

```
FAILED tests/test_intl.py::test_gettext_dont_rebuild_mo - AssertionError: assert 1 == 0
 +  where 1 = <function test_gettext_dont_rebuild_mo.<locals>.get_number_of_update_targets at 0x0000022947B2A8E0>(<SphinxTestApp buildername='dummy'>)
= 1 failed, 1787 passed, 23 skipped, 5 xfailed, 28 xpassed, 6 warnings in 271.50s (0:04:31) =
Error: Process completed with exit code 1.
```

which I have observed a number of times recently either on PRs or direct pushes to master.

### How to Reproduce

See https://github.com/sphinx-doc/sphinx/actions/runs/4383723023/jobs/7674313285 or earlier failed test of #11224 prior to merge to master: https://github.com/sphinx-doc/sphinx/actions/runs/4354037120/jobs/7608843307

As this seems random, I have no recipe for reproducing.

### Environment Information

```text
not relevant
```

**Comment by jayaddison** [comment]
So:

- The test uses the ['dummy' builder](https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/builders/dummy.py), that does not emit any output (great, that helps narrow down the cause)
- The most recent assertion failure occurred during `phase1` of the test, before any files are removed and recreated (also good in terms of narrowing the cause)
- The assertion is on the number of modified/changed files according to [`env.get_outdated_files`](https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/environment/__init__.py#L454-L509)
- There are six locations within that method that call `changed.add(...)` -- these are the potential source(s)
  - I think that the first two of those can be ignored because the test uses the dummy builder and doesn't do anything to configure `reread_always`
  - [Another one](https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/environment/__init__.py#L495-L499) can be ignored, again because the builder is not writing to the target directory.

The remaining items contain two `mtime`-related modification checks, and one `OSError` error handler case.

I think it's unlikely that the `OSError` case is the cause (although it might be better to handle it by adding the relevant files to a new, fourth output tuple entry -- `unknown` -- rather than bundling them into `changed`).

In summary:

My best guess at the moment is that we're running into a filesystem timestamp consistency issue on some of the runners:

https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/builders/__init__.py#L498-L503

And in terms of how to fix that, for test stability purposes:

Not sure yet.

**Comment by jayaddison** [comment]
It's tricky without having a way to replicate the problem reliably, but I'd like to try reverting the `... = max(time.time(), path.getmtime(...))` logic (718993afc8e0e61c689a7681aa27f0f6dfbbcb7e) and going back to the simpler `... = time.time()` logic that preceded it.  I'm repeat-CI-testing that in a bundle of changes in jayaddison/sphinx#3 at the moment.

I've read one or two bugs on the cPython bugtracker that seem to indicate that Windows timestamp granularity had some issues and has been improved since 718993afc8e0e61c689a7681aa27f0f6dfbbcb7e was introduced - and also I think it's _probably_ relatively rare for people to be building Sphinx documentation on low-timestamp-granularity filesystems (particularly FAT32).  So I'm thinking of it as a potential tradeoff between unit test stability and (perhaps) some rare edge cases.

**Comment by jayaddison** [comment]
> It's tricky without having a way to replicate the problem reliably, but I'd like to try reverting the `... = max(time.time(), path.getmtime(...))` logic ([718993a](https://github.com/sphinx-doc/sphinx/commit/718993afc8e0e61c689a7681aa27f0f6dfbbcb7e)) and going back to the simpler `... = time.time()` logic that preceded it. I'm repeat-CI-testing that in a bundle of changes in [jayaddison#3](https://github.com/jayaddison/sphinx/pull/3) at the moment.

Update: reverting 718993afc8e0e61c689a7681aa27f0f6dfbbcb7e _does not_ fix the problem.  For a situation where that commit is reverted and yet the test still fails (after a few previous successful test results), see here: https://github.com/jayaddison/sphinx/actions/runs/5038386691/jobs/9035841524?pr=3#step:5:1931

**Comment by jayaddison** [comment]
I've [added some debug printout code](https://github.com/jayaddison/sphinx/pull/3/commits/68dd1548d22bf18eb89dbf4e0985243bcc10ce8d), and here is the output from it during a [recent unit test run](https://github.com/jayaddison/sphinx/actions/runs/5038672907/jobs/9036325271?pr=3#step:5:1931) where the failure occurred:

```
===================
docname:  index
===================
mtime:    1684687126.3444436
newmtime: 1684686899.1642947
changed?  False

===================
docname:  bom
===================
mtime:    1684687126.3444436
newmtime: 1684686899.1642947
changed?  False

-------------------
deppath:  C:\Users\runneradmin\AppData\Local\Temp\pytest-of-runneradmin\pytest-0\builder-gettext-dont-rebuild-mo\.\xx\LC_MESSAGES\bom.mo
-------------------
mtime:    1684687126.3444436
depmtime: 1684687126.3444438
changed?  True
```

**Comment by jayaddison** [comment]
This is based on a small sample size so far, but [switching to use integer nanosecond-precision timestamps](https://github.com/jayaddison/sphinx/pull/3/commits/96359aac8bde47fa34142d6cdff46469dcd595c7) instead of the existing floating-point second-precision timestamps may be helping.

If that remains true after a few more rounds of testing, then I'll open a fix pr for this issue, containing the revert of 718993afc8e0e61c689a7681aa27f0f6dfbbcb7e and the nanosecond-timestamp switchover.

**Comment by jayaddison** [comment]
Let's re-open this, because the problem [still seems to occur after #11435 was merged](https://github.com/sphinx-doc/sphinx/pull/11435#issuecomment-1644583559).

**Comment by jayaddison** [comment]
Because the [additional informational logging added for this test failure](https://github.com/sphinx-doc/sphinx/pull/11500) [points to the byte-order-mark (`bom`) file as a source of the problem](https://github.com/sphinx-doc/sphinx/actions/runs/5615685910/job/15216466568#step:5:2007), I'm wondering about [these unit test code lines](https://github.com/sphinx-doc/sphinx/blob/71db08c05197545944949d5aa76cd340e7143627/tests/test_intl.py#L58-L60) that rewrite that file under some circumstances.

Perhaps adding further information to indicate whether that test code logic is evaluated could confirm whether it has a causal relationship with the test failures.

**Comment by jayaddison** [comment]
What do you think about adding a debug print [between these lines](https://github.com/sphinx-doc/sphinx/blob/71db08c05197545944949d5aa76cd340e7143627/tests/test_intl.py#L59-L60) to determine whether that's a potential cause, @AA-Turner?  (I've drafted a commit for that at https://github.com/jayaddison/sphinx/commit/5924a4db882ee01d9f206a33417f55ec95118b12 )

**Comment by jayaddison** [comment]
Taking another look at this currently; I'm going to repeat-run CI on my fork with 5924a4db882ee01d9f206a33417f55ec95118b12 in place to see whether that helps to track anything down.

**Comment by jayaddison** [comment]
[This build job failed](https://github.com/jayaddison/sphinx/actions/runs/6186279550/job/16793535975) and the additional output was:

```
# compiling .mo file C:\Users\runneradmin\AppData\Local\Temp\pytest-of-runneradmin\pytest-0\builder-gettext-dont-rebuild-mo\xx\LC_MESSAGES\xx\LC_MESSAGES\bom.mo
```

I'm not sure yet whether that helps trace this down.  I'm trying to refresh my context about how all this works.

**Comment by jayaddison** [comment]
Does anyone have any thoughts on whether we could/should mark this test as `xfail` on non-posix platforms and reinvestigate at a later date?

**Comment by AA-Turner** [comment]
We have strict xfail turned on, so that would introduce stochastic failures in the other direction, unhelpfully.

A

**Comment by jayaddison** [comment]
Ok - although that only changes the default, I think - so we could specify `strict=False` in the marker.

I'll admit to likely truth of: it's the kind of bug/issue that once hidden, might not really be fixed later (or not until 'much, much' later at least).  (subtle, possibly complicated, potentially time-consuming to track down, and seemingly low-impact in terms of production usage (although arguably important for continuous integration reporting))

**Comment by picnixz** [comment]
Fixed in #11940
